### PR TITLE
copy doc output files by format

### LINKED ISSUE #103785: `x doc std` unconditionally recompiles the standard library each time
I tried this:

```sh
x doc --stage 0 std
x doc --stage 0 std
```

I expected to see this happen: The second invocation takes essentially no time.

Instead, this happened: The docs were completely rebuilt.

The problem is that we remove the docs between each invocation: https://github.com/rust-lang/rust/blob/c2b0cc14d1da274cb8548aff5030dd30b16f67e1/src/bootstrap/doc.rs#L613-L615

@lukemathwalker this is a regression from https://github.com/rust-lang/rust/pull/101799 - do you think you'll have time to follow up? I think we can either add a stamp file to avoid rebuilding, or find some other way to avoid removing the whole directory - maybe we can only remove/copy the `.json` files, rather than everything?

### Meta

HEAD is branched from 77e57db384aca99444c3b5f6a9c86bc58a804d89.

### PR DESCRIPTION
**Pull request by onur-ozkan**
This pr provides copying doc outputs by checking output format without removing output directory on each trigger.

Resolves #103785

### CODE REVIEW

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -551,6 +551,53 @@ fn doc_std(
...
+
+                match format {
+                    DocumentationFormat::HTML if extension != Some("json") => {
+                        builder.copy(&path, &dst)
+                    }
+                    DocumentationFormat::JSON
+                        if extension == Some("json") || name.to_str() == Some(".stamp") =>
+                    {
+                        builder.copy(&path, &dst)
+                    }
```
Comment (ozkanonur) [inline src/bootstrap/doc.rs:594]: I am not sure if this conditions will be enough for the purpose. What do you think ? @jyn514

**Comment by ozkanonur** [comment]
pr is ready for review @Mark-Simulacrum @jyn514

**Comment by Mark-Simulacrum** [comment]
Is there a reason we need to have a shared output directory for our invocations of cargo doc for JSON and not JSON? That seems much less picky in terms of picking file types out by extension...

In particular, it seems like the HTML format has some JSON files so the approach of this PR won't quite work:

```
~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share$ find doc/rust/html | grep json
doc/rust/html/book/searchindex.json
doc/rust/html/book/second-edition/searchindex.json
doc/rust/html/book/2018-edition/searchindex.json
doc/rust/html/book/first-edition/searchindex.json
doc/rust/html/rust-by-example/searchindex.json
doc/rust/html/error_codes/searchindex.json
doc/rust/html/cargo/searchindex.json
doc/rust/html/clippy/searchindex.json
doc/rust/html/rustc/searchindex.json
doc/rust/html/rustc/json.html
doc/rust/html/embedded-book/searchindex.json
doc/rust/html/rustdoc/searchindex.json
doc/rust/html/nomicon/searchindex.json
doc/rust/html/src/test/formatters/json.rs.html
doc/rust/html/unstable-book/searchindex.json
doc/rust/html/reference/searchindex.json
doc/rust/html/edition-guide/searchindex.json
```

**Comment by jyn514** [comment]
Hmm, won't it hurt caching if cargo has to rebuild the dependencies in between? We can only control the target directory cargo uses, not the `--out-dir` it passes to rustdoc.

I agree this seems to cause issues though ... maybe a stamp file is better after all.

**Comment by ozkanonur** [comment]
> In particular, it seems like the HTML format has some JSON files so the approach of this PR won't quite work:
> 
> ```
> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/share$ find doc/rust/html | grep json
> doc/rust/html/book/searchindex.json
> doc/rust/html/book/second-edition/searchindex.json
> doc/rust/html/book/2018-edition/searchindex.json
> doc/rust/html/book/first-edition/searchindex.json
> doc/rust/html/rust-by-example/searchindex.json
> doc/rust/html/error_codes/searchindex.json
> doc/rust/html/cargo/searchindex.json
> doc/rust/html/clippy/searchindex.json
> doc/rust/html/rustc/searchindex.json
> doc/rust/html/rustc/json.html
> doc/rust/html/embedded-book/searchindex.json
> doc/rust/html/rustdoc/searchindex.json
> doc/rust/html/nomicon/searchindex.json
> doc/rust/html/src/test/formatters/json.rs.html
> doc/rust/html/unstable-book/searchindex.json
> doc/rust/html/reference/searchindex.json
> doc/rust/html/edition-guide/searchindex.json
> ```

May I  know how did you achieve this?

**Comment by Mark-Simulacrum** [comment]
The files in question can be installed with `rustup +stable component add rust-docs` and then the command and where I ran it from is in the snippet.

**Comment by Mark-Simulacrum** [comment]
> Hmm, won't it hurt caching if cargo has to rebuild the dependencies in between?

I think in practice re-building the standard library and its dependencies shouldn't take long enough to be worried about it for now -- and if we follow the suggestion around different destinations via symlink updates or similar (below) then it might not need to rebuild dependencies.

> We can only control the target directory cargo uses, not the --out-dir it passes to rustdoc.

I think we do some symlink'ing and such to adjust *where* the out-dir passed to rustdoc actually ends up, fwiw, but it's not exactly the most ideal thing.

**Comment by ozkanonur** [comment]
> I think in practice re-building the standard library and its dependencies shouldn't take long enough to be worried about it for now -- and if we follow the suggestion around different destinations via symlink updates or similar (below) then it might not need to rebuild dependencies.

Should I work on fixing the issue you mentioned then?

**Comment by jyn514** [comment]
@ozkanonur yes please, if you use separate `--target-dir` flags for JSON and non-json output that should sidestep this issue altogether (and we can worry about the compile time later, I think Mark is right it won't be a big deal in practice).

**Comment by ozkanonur** [comment]
> We can only control the target directory cargo uses, not the `--out-dir` it passes to rustdoc.

I tried to check if I can play with `--out-dir` by checking doc format, however I couldn't figure how `-o` is passed to here https://github.com/rust-lang/rust/blob/e07425d55b77fde99af2092a92109a0da0860692/src/librustdoc/lib.rs#L705

At least now I have more experience about the rustdoc process pipelines.

> if you use separate --target-dir flags for JSON and non-json output that should sidestep this issue altogether

With the last commit I aplied it for std doc(I think we don't need this for compiler-docs since it's always including html docs. Please correct me if I am wrong.).

**Pull request merged**